File: bridge/DiscoveredExtensionInjector.ts
import { ExtensionInjector } from "./ExtensionInjector.js";

export class DiscoveredExtensionInjector extends ExtensionInjector {
  async inject() {
    const discovered = await this.discoverReadyServiceWorker();
    if (discovered) return { ...discovered, source: "discovered" };
    if (!this.options.require_service_worker_target) return null;
    const waited = await this.waitForReadyServiceWorker(this.options.service_worker_ready_timeout_ms ?? 60_000, {
      matched_only: this.options.trust_matched_service_worker,
    });
    if (waited) return { ...waited, source: "discovered" };
    throw new Error(
      `Required ModCDP service worker target was not visible ` +
        `(${[
          ...(this.options.service_worker_url_includes ?? []),
          ...(this.options.service_worker_url_suffixes ?? []),
        ].join(", ") || "no matcher"}).`,
    );
  }
}

File: client/python/modcdp/DiscoveredExtensionInjector.py
from __future__ import annotations

from .ExtensionInjector import DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS, DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS, ExtensionInjector, ExtensionInjectionResult


class DiscoveredExtensionInjector(ExtensionInjector):
    def inject(self) -> ExtensionInjectionResult | None:
        discovered = self.discoverReadyServiceWorker()
        if discovered:
            return {**discovered, "source": "discovered"}
        if self.options.get("trust_matched_service_worker"):
            waited = self.waitForReadyServiceWorker(
                self.options.get("service_worker_probe_timeout_ms") or DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
                matched_only=True,
            )
            if waited:
                return {**waited, "source": "discovered"}
        woke = self.wakeConfiguredExtension()
        if woke:
            waited = self.waitForReadyServiceWorker(
                self.options.get("service_worker_probe_timeout_ms") or DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
                matched_only=bool(self.options.get("trust_matched_service_worker")),
            )
            if waited:
                return {**waited, "source": "discovered"}
        if not self.options.get("require_service_worker_target"):
            return None
        waited = self.waitForReadyServiceWorker(
            self.options.get("service_worker_ready_timeout_ms") or DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS,
            matched_only=bool(self.options.get("trust_matched_service_worker")),
        )
        if waited:
            return {**waited, "source": "discovered"}
        matchers = ", ".join(
            [
                *(self.options.get("service_worker_url_includes") or []),
                *(self.options.get("service_worker_url_suffixes") or []),
            ]
        )
        raise RuntimeError(f"Required ModCDP service worker target was not visible ({matchers or 'no matcher'}).")
